crates/astria-core/src/sequencerblock/v1/block/mod.rs

```diff
@@ -1165,122 +1162,424 @@ impl ExtendedCommitInfoError {
+    ///
+    /// Returns an error if the data is not in this order, or cannot be parsed.
+    #[expect(clippy::missing_panics_doc, reason = "can't panic")]
+    #[expect(
+        clippy::too_many_lines,
+        reason = "code wouldn't benefit overall from being split into smaller functions"
+    )]
+    pub fn new_from_typed_data(
+        data: &[Bytes],
+        with_extended_commit_info: bool,
+    ) -> Result<Self, SequencerBlockError> {
+        let mut merkle_tree_leaves = Vec::with_capacity(data.len());
+
+        let mut data_iter = data.iter().enumerate().peekable();
+
+        // First item must be a rollup txs root.
+        let (rollup_txs_root_index, bytes) = data_iter
+            .next()
+            .ok_or(SequencerBlockError::no_rollup_transactions_root())?;
+        let rollup_transactions_root = DataItem::decode(rollup_txs_root_index, bytes.as_ref())
+            .and_then(|data_item| data_item.into_rollup_transactions_root(rollup_txs_root_index))
+            .map_err(SequencerBlockError::data_item)?;
+        merkle_tree_leaves.push(Sha256::digest(rollup_transactions_root));
+
+        // Second item must be a rollup IDs root.
+        let (rollup_ids_root_index, bytes) = data_iter
+            .next()
+            .ok_or(SequencerBlockError::no_rollup_ids_root())?;
+        let rollup_ids_root = DataItem::decode(rollup_ids_root_index, bytes.as_ref())
+            .and_then(|data_item| data_item.into_rollup_ids_root(rollup_ids_root_index))
+            .map_err(SequencerBlockError::data_item)?;
+        merkle_tree_leaves.push(Sha256::digest(rollup_ids_root));
+
+        // Third item might be upgrade change hashes - peek to check before advancing the iterator.
+        let maybe_upgrade_item = matches!(
+            data_iter.peek().map(|(index, bytes)| {
+                DataItem::decode(*index, bytes.as_ref()).map_err(SequencerBlockError::data_item)
+            }),
+            Some(Ok(DataItem::UpgradeChangeHashes(_)))
+        )
+        .then(|| {
+            // The next item is upgrade change hashes - advance the iterator and parse the item.
+            let (index, bytes) = data_iter
+                .next()
+                .expect("must be `Some`, since it's `Some` in peek");
+            let upgrade_change_hashes = DataItem::decode(index, bytes.as_ref())
+                .and_then(|data_item| data_item.into_upgrade_change_hashes(index))
+                .expect("must be upgrade, since it's upgrade in peek");
+            // Concatenate the raw bytes of all the change hashes to use as the input for the
+            // Merkle tree leaf for this item.
+            let concatenated_hashes = concatenate_upgrade_change_hashes(&upgrade_change_hashes);
+            merkle_tree_leaves.push(Sha256::digest(&concatenated_hashes));
+            Ok((index, upgrade_change_hashes))
+        })
+        .transpose()?;
+
+        // If `with_extended_commit_info` is true, the next item must be an extended commit info.
```

Comment:
do we still need `with_extended_commit_info`? if we're shipping oracle and `DataItem` in upgrade1, can't we assume if this function is called, there always will be extended commit info?

Reply:
i guess the only block where this isn't the case is the block at which the vote extensions are enabled, since there won't be extended commit info until the next block.

Reply:
There's that case, yes, but also we still support a new network which doesn't enable the price feed and extended commit info.